CF-609E C++ Solution (#2789)
cpinitiative/usaco-guide: solutions/cf-609E.mdx

---
id: cf-609E
source: CF
title: Minimum spanning tree for each edge
author: Chongtian Ma
---

[Official Analysis](https://codeforces.com/blog/entry/22187)

## Implementation 

As mentioned in the official editorial, the answer for each edge is removing the heaviest edge in the MST in
the path from vertex $u$ to $v$ and add the given weight. This will ensure that the MST is still connected.

**Time Complexity:** $\mathcal{O}(M\log N)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

> const int MAXN = 2e5 + 1;

Reviewer: MAXN is really 2e5, so either 0-index or use MAXN + 1 where needed instead

Author: well its the MAXN if one indexed, I can add a comment but i think using MAXN + 1 everywhere will be more confusing?

Reviewer: or you could just, yknow, 0-index

Author: but ill be adding one to more things and things become uncool

Reviewer: just sub 1 from the node indices?

const int MAXL = 20; // approximately maximum log

// for euler-tour
array<int, MAXN> enter_time, exit_time, depth;

// for binary lifting
array<array<int, MAXL>, MAXN> up, max_up;
// up[i][j] stores the 2^jth ancestor of vertex i
// max_up[i][j] stores the maximum weight of all the edges

> // from i to its 2^jth ancestor

Reviewer: 1. put these above the variable declarations
2. separate lines for var declarations

map<pair<int, int>, int> weight; // weight of edge {i, j}
vector<int> graph[MAXN]; // adjacency list for the original MST
int timer = 1;
int n, m;

// here we build the MST using kruskals' algorithm
// returns the sum of the weights that took to build the MST
ll kruskal(vector<array<int,4>>& edges){
	
	// sort by non-decreasing weight
	sort(edges.begin(), edges.end(), [&](auto a, auto b){
		return a[3] < b[3];
	});
	
	// implementation of basic Disjoint Set Union
	vector<int> parent(n+1), sz(n+1, 1);

> 	iota(parent.begin(), parent.end(), 0);

Reviewer: iota? if this does what i think it does, c++ already fills the vector with 0

Author: search up the cpp docs

			return true;
		}
		return false;
	};
	
	ll total_cost = 0;
	for(array<int, 4> edge : edges){
		int u = edge[1], v = edge[2], w = edge[3];
		if(unite(u, v)){
			total_cost += w;
			graph[u].push_back(v);
			graph[v].push_back(u);
		}
	}
	
	return total_cost;
	
}

void euler_tour(int vertex, int parent){
	enter_time[vertex] = timer++;
	depth[vertex] = depth[parent] + 1;
	
	// initialize binary lifting arrays
	up[vertex][0] = parent; 
	max_up[vertex][0] = weight[{parent, vertex}];
	
	for(int child: graph[vertex]){
		if(child != parent){
			euler_tour(child, vertex);
		}
	}
	
	exit_time[vertex] = timer - 1;
}

bool is_ancestor(int x, int y){
	return enter_time[x] <= enter_time[y] && exit_time[y] <= exit_time[x];
}

int lca(int x, int y){
	if(is_ancestor(x, y)){
		return x;
	}
	for(int i = MAXL-1; i >= 0; i--){
		if(!is_ancestor(up[x][i], y)){
			x = up[x][i];
		}
	}
	return up[x][0];
}

// get maximum weight of all the edges from x to its kth ancestor
int getmaxup(int x, int k){
	int res = 0;
	for(int i = 0; i < MAXL; i++){
		if(k & (1 << i)){
			res = max(res, max_up[x][i]);
			x = up[x][i];
		}
	}
	return res;
}

int main(){
	ios::sync_with_stdio(false);
	cin.tie(0);

> 	

Reviewer: remove if this runs w/o

	cin >> n >> m;
	
	vector<array<int, 4>> edges(m); // {index, from, to, weight}
	for(int i = 0; i < m; i++){
		int u, v, w; 
		cin >> u >> v >> w;
		edges[i] = {i, u, v, w};
		weight[{u, v}] = w;
		weight[{v, u}] = w;
	}
	ll total_cost = kruskal(edges);
	
	euler_tour(1, 1);
	// fill binary-lifting arrays
	for(int k = 1; k < MAXL; k++){
		for(int i = 1; i <= n; i++){
			up[i][k] = up[up[i][k - 1]][k - 1];
			// take max of weights from left and right
			max_up[i][k] = max(max_up[i][k - 1], max_up[up[i][k - 1]][k - 1]);
		}
	}
	
	vector<ll> ans(m);

> 	for(array<int, 4> edge : edges){

Reviewer: maybe use a struct for the edges

Author: i explained how the array works in the main function. is it necessary?

Reviewer: or just use a struct with meaningful names

		int index = edge[0], u = edge[1], v = edge[2], w = edge[3];
		int least_common_ancestor = lca(u, v);
		int max_u = getmaxup(u, depth[u] - depth[least_common_ancestor]);
		int max_v = getmaxup(v, depth[v] - depth[least_common_ancestor]);
		ans[index] = total_cost - max(max_u, max_v) + w;
	}
	

> 	for(int i = 0; i < m; i++){

Reviewer: spaces for control flow statements

Author: example?

Reviewer: ```suggestion
	for (int i = 0; i < m; i++) {
```

---
id: cf-609E
source: CF

> title: Minimum spanning tree for each edge

Reviewer 2: not the correct title

author: Chongtian Ma
---

> [Official Analysis](https://codeforces.com/blog/entry/22187)

Reviewer 2: add lang

Author: the editorial doesnt provide a sample code...

Reviewer 2: oops ok mb

## Implementation 

As mentioned in the official editorial, the answer for each edge is removing the heaviest edge in the MST in

> the path from vertex $u$ to $v$ and add the given weight. This will ensure that the MST is still connected.

Reviewer 2: if it's already mentioned in the editorial, why mention it here?

at least add some extra intuition/insight

also move out of implementation

Author: is it better now?

**Time Complexity:** $\mathcal{O}(M\log N)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

const int MAXN = 2e5;

> const int MAXL = 20; // approximately maximum log

Reviewer 2: two spaces for inline comments

// for euler-tour
array<int, MAXN> enter_time, exit_time, depth;

// up[i][j] stores the 2^jth ancestor of vertex i (for binary lifting)
array<array<int, MAXL>, MAXN> up;

/*
 * max_up[i][j] stores the maximum weight of all the edges
 * from i to its 2^jth ancestor
 */ 
array<array<int, MAXL>, MAXN> max_up;

map<pair<int, int>, int> weights;  // weight of edge {i, j}
vector<int> graph[MAXN];  // adjacency list for the original MST
int timer = 1;
int n, m;

struct Edge {
	// bidirectional edge from u to v
	int index, u, v, weight;
};

// here we build the MST using kruskals' algorithm
// it returns the sum of the weights that took to build the MST 
ll kruskal(vector<Edge>& edges) {
	
	// sort by non-decreasing weight
	sort (edges.begin(), edges.end(), [&](Edge a, Edge b) {
		return a.weight < b.weight;
	});
	
	// implementation of basic Disjoint Set Union
	vector<int> parent(n), sz(n, 1);
	iota(parent.begin(), parent.end(), 0);
	
	function<int(int)> get = [&](int x) {
		if (parent[x] != x) {
			parent[x] = get(parent[x]);
		}
		return parent[x];
	};
	

> 	// returns true if both vertexes are not already united

Reviewer 3: ```suggestion
	// returns true if the vertices are not already united
```

	function<bool(int, int)> unite = [&](int x, int y) {
		int parent_x = get(x), parent_y = get(y);
		if (sz[parent_x] < sz[parent_y]) {
			swap(parent_x, parent_y);
		}
		if (parent_x != parent_y) {
			parent[parent_y] = parent_x;
			sz[parent_x] += sz[parent_y];
			return true;
		}
		return false;
	};
	
	ll total_cost = 0;
	for (Edge e : edges) {
		if (unite(e.u, e.v)) {
			total_cost += e.weight;
			graph[e.u].push_back(e.v);
			graph[e.v].push_back(e.u);
		}
	}
	
	return total_cost;
	
}

void euler_tour(int vertex, int parent) {
	enter_time[vertex] = timer++;
	depth[vertex] = depth[parent] + 1;
	
	// initialize binary lifting arrays
	up[vertex][0] = parent; 
	max_up[vertex][0] = weights[{parent, vertex}];
	
	for (int child: graph[vertex]) {
		if (child != parent) {
			euler_tour(child, vertex);
		}
	}
	
	exit_time[vertex] = timer - 1;
}

bool is_ancestor(int x, int y) {
	return enter_time[x] <= enter_time[y] && exit_time[y] <= exit_time[x];
}

int lca(int x, int y) {
	if (is_ancestor(x, y)) {
		return x;
	}

> 	for (int i = MAXL-1; i >= 0; i--) {

Reviewer 3: ```suggestion
	for (int i = MAXL - 1; i >= 0; i--) {
```

		if (!is_ancestor(up[x][i], y)) {
			x = up[x][i];
		}
	}
	return up[x][0];
}

// get maximum weight of all the edges from x to its kth ancestor

> int getmaxup(int x, int k) {

Reviewer 3: (optional) rename function to `get_max_up` to match the other functions

	int res = 0;
	for (int i = 0; i < MAXL; i++) {
		if (k & (1 << i)) {
			res = max(res, max_up[x][i]);
			x = up[x][i];
		}
	}
	return res;
}

int main() {
	cin >> n >> m;
	
	vector<Edge> edges(m);
	for (int i = 0; i < m; i++) {
		int u, v, w; 
		cin >> u >> v >> w;
		edges[i] = {i, --u, --v, w};
		weights[{u, v}] = w;
		weights[{v, u}] = w;
	}
	ll total_cost = kruskal(edges);
	
	euler_tour(0, 0);

> 	// fill binary-lifting arrays

Reviewer 2: ```suggestion
	
	// fill binary-lifting arrays
```

Reviewer: (optional imo)

---
id: cf-609E
source: CF
title: Minimum spanning tree for each edge
author: Chongtian Ma
---

[Official Analysis](https://codeforces.com/blog/entry/22187)

## Implementation 

As mentioned in the official editorial, the answer for each edge is removing the heaviest edge in the MST in
the path from vertex $u$ to $v$ and add the given weight. This will ensure that the MST is still connected.

**Time Complexity:** $\mathcal{O}(M\log N)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

const int MAXN = 2e5;
const int MAXL = 20; // approximately maximum log

// for euler-tour
array<int, MAXN> enter_time, exit_time, depth;

> // for binary lifting:
> // up[i][j] stores the 2^jth ancestor of vertex i

Reviewer: multiline comment or just
```suggestion
// up[i][j] stores the 2^jth ancestor of vertex i (for binary lifting)
```

array<array<int, MAXL>, MAXN> up;

/*
 * max_up[i][j] stores the maximum weight of all the edges
 * from i to its 2^jth ancestor
 */ 
array<array<int, MAXL>, MAXN> max_up;

> map<pair<int, int>, int> weights; // weight of edge {i, j}

Reviewer: two spaces after code before starting inline comment
```suggestion
map<pair<int, int>, int> weights;  // weight of edge {i, j}
```

vector<int> graph[MAXN]; // adjacency list for the original MST
int timer = 1;
int n, m;

struct Edge {
	// bidirectional edge from u to v
	int index, u, v, weight;
};

// here we build the MST using kruskals' algorithm

> // it returns the sum of the weights that took to build the MST 

Reviewer: multiline comment here as well

Author: its two different statements though

using namespace std;
using ll = long long;

const int MAXN = 2e5;
const int MAXL = 20; // approximately maximum log

// for euler-tour
array<int, MAXN> enter_time, exit_time, depth;

// for binary lifting:
// up[i][j] stores the 2^jth ancestor of vertex i
array<array<int, MAXL>, MAXN> up;

// max_up[i][j] stores the maximum weight of all the edges

> // from i to its 2^jth ancestor

Reviewer: ```cpp
/*
 * multiline comments are done
 * like this
 */
```

array<array<int, MAXL>, MAXN> max_up;

map<pair<int, int>, int> weights; // weight of edge {i, j}
vector<int> graph[MAXN]; // adjacency list for the original MST
int timer = 1;
int n, m;

> struct edge {

Reviewer: rename to `Edge`
